### PR TITLE
feat(eslint): Enable no-non-null-assertion rule

### PR DESCRIPTION
**Pull request by gkartalis**
This PR resolves [] 

### Description

- [x] Enables `no-non-null-assertion` rule in eigen but **excludes tests**

The reasoning behind enabling this is that this non-safe way of writing code makes the app crash and in eigen specifically it triggers many unable to load screens.

There are around ~600 instances of this in `eigen` right now. Will not complain if you are writing code unless you touch a file that has one of these errors.

I believe it is better to address them like this but happy to hear your thoughts / concerns in case you dissagree.

### More info / Usage

TypeScript's `!` non-null assertion operator asserts to the type system that an expression is non-nullable, as in not `null` or `undefined`.
Using assertions to tell the type system new information is often a sign that code is not fully type-safe.
It's generally better to structure program logic so that TypeScript understands when values may be nullable.

#### ❌ Incorrect

```ts
interface Example {
  property?: string;
}

declare const example: Example;
const includesBaz = example.property!.includes('baz');
```

#### ✅ Correct

```ts
interface Example {
  property?: string;
}

declare const example: Example;
const includesBaz = example.property?.includes('baz') ?? false;
```

### PR Checklist

- [x] I have tested my changes on **iOS** and **Android**.
- [x] I hid my changes behind a **[feature flag]**, or they don't need one.
- [x] I have included **screenshots** or **videos**, or I have not changed the UI.
- [x] I have added **tests**, or my changes don't require any.
- [x] I added an **[app state migration]**, or my changes do not require one.
- [x] I have documented any **follow-up work** that this PR will require, or it does not require any.
- [x] I have added a **changelog entry** below, or my changes do not require one.

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to **run** this PR on the simulator or device.

<details><summary>Changelog updates</summary>

#### Cross-platform user-facing changes

-

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

- lint - enable no-non-null-assertion rule - gkartalis

</details>

Need help with something? Have a look at our [docs], or get in touch with us.

[app state migration]: ../blob/main/docs/adding_state_migrations.md
[feature flag]: ../blob/main/docs/developing_a_feature.md
[docs]: ../blob/main/docs/README.md

### CODE REVIEW

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  ❌ **An error occurred while validating your changelog, please make sure you provided a valid changelog.**
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 2b119642d1a72105d49ba889e8965baa8234afda
</p>

File: src/app/Scenes/Articles/Articles.tests.tsx [hunk 1]
```diff
@@ -10,7 +10,7 @@ describe("Articles", () => {
   it("renders articles", async () => {
     renderWithRelay({
       Query: () => ({
-        articlesConnection: mockArticlesConnection,
+        articlesConnection: mockArticlesConnection!,
```
Comment (gkartalis) [inline src/app/Scenes/Articles/Articles.tests.tsx:13]:
```suggestion
        articlesConnection: mockArticlesConnection,
```

**Review by araujobarret — APPROVED** [review]

**Review by olerichter00 — APPROVED** [review]
🍃 🌟 🍃

**Review by damassi — DISMISSED** [review]

**Review by MounirDhahri — APPROVED** [review]
Nice!

**Pull request merged**